compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/DelegationLowering.kt

```diff
@@ -0,0 +1,236 @@
+    description = "Construct KProperty instances to be used when accessing delegated properties"
+)
+
+internal class PropertyDelegationLowering(val context: JvmBackendContext) : ClassLoweringPass {
+    private object DECLARATION_ORIGIN_KPROPERTIES_FOR_DELEGATION :
+        IrDeclarationOriginImpl("KPROPERTIES_FOR_DELEGATION", true)
+
+    private val plainJavaClass =
+        context.getIrClass(FqName("java.lang.Class")).owner
+
+    private val reflectionClass =
+        context.getIrClass(FqName("kotlin.jvm.internal.Reflection")).owner
+
+    private val getOrCreateKotlinClass =
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier("getOrCreateKotlinClass") }
+
+    private val getOrCreateKotlinPackage =
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier("getOrCreateKotlinPackage") }
+
+    private fun propertyTypeAndFactory(prefix: String, i: Int) =
+        context.getIrClass(FqName("kotlin.jvm.internal.${prefix.capitalize()}Reference${i}Impl")) to
+                reflectionClass.declarations.find { it.descriptor.name == Name.identifier("${prefix}${i}") }
+
+    private val properties =
+        arrayOf(0, 1, 2).map { propertyTypeAndFactory("property", it) }
+
+    private val mutableProperties =
+        arrayOf(0, 1, 2).map { propertyTypeAndFactory("mutableProperty", it) }
+
+    private val arrayItemGetter =
+        context.ir.symbols.array.owner.functions.single { it.name == Name.identifier("get") }
+
+    private val kPropertyType =
+        context.reflectionTypes.kProperty.toIrType()
+
+    private val kPropertiesFieldType =
+        context.ir.symbols.array.createType(false, listOf(makeTypeProjection(kPropertyType, Variance.OUT_VARIANCE)))
+
+    override fun lower(irClass: IrClass) {
+        // Maintain an array of `KProperty` instances, one per delegated property, as the delegate target is allowed to use identity
+        // comparison of these references for differentiating properties.
+        val kProperties = mutableMapOf<CallableDescriptor, Pair<IrExpression, Int>>()
+        val kPropertiesField = WrappedFieldDescriptor(Annotations.create(listOf())).let { descriptor ->
+            IrFieldImpl(
+                UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+                DECLARATION_ORIGIN_KPROPERTIES_FOR_DELEGATION,
+                IrFieldSymbolImpl(descriptor),
+                "\$delegatedProperties".synthesizedName,
+                kPropertiesFieldType,
+                Visibilities.PRIVATE,
+                isFinal = true,
+                isExternal = false,
+                isStatic = true
+            ).apply { descriptor.bind(this) }
+        }
+
+        irClass.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
+            override fun visitPropertyReference(expression: IrPropertyReference): IrExpression =
+                // This actually generates KProperties for all property references, not just delegated ones.
```

Comment:
Property references should not be processed as delegated ones: anonymous class for them should be generated in CallableReferenceLowering.kt to support non-reflecsive get/set/invoke

Reply:
That makes sense, but I don't understand why not do that for delegated references as well?

Reply:
Do you mean this case:
```
val z by lazy  { 1 } // propertyMetadata

fun main() {
    ::z.get()     //new anonymous class  should be generated
}
``` ?

Reply:
As I understand it, a delegated property

```
class C {
    val x by Delegate()
}
```

is simply transformed into something that resembles this:

```
class C {
    val x$delegate = Delegate()
    val x
        get() = x$delegate.getValue(this, C::x)
        set(val: T) = x$delegate.setValue(this, C::x, val)
}
```

and the code in this lowering is required to generate an appropriate KProperty in place of `C::x`. So my question is, why handle `C::x` here and a reference to a non-delegated property elsewhere differently? Seems to me like both can use a subclass of KProperty in which overrides `get` and `set` call `C::getX` and `C::setX` directly instead of going through reflection. (Of course, calling `get`/`set` on a KProperty passed to `getValue`/`setValue` is of rather limited usefulness, so the performance of that might not matter, but still.)

Reply:
> Of course, calling get/set on a KProperty passed to getValue/setValue is of rather limited usefulness, so the performance of that might not matter, but still.

Yes, it's useless (and in most case will produce SOE)
Reflection here is used to avoid a lot of anonymous class generation.

Please also note that there are two strategy for local delegation properties: 
```
object O {
    operator fun getValue(a: Any?, b: Any?): String = "OK"
}

fun noInline() {
    val z by O //same as z$delegate.getValue(this, ::z)
}

inline fun InlineFoo() {
    val b by O // after inline access to $$delegatedProperties in most cases are lost so anonymous object as for general CallableReference is generated
}
```